src/dataflow/components/dataflow-program.tsx

```diff
@@ -450,22 +453,34 @@ export class DataflowProgram extends BaseComponent<IProps, IState> {
 
   private hasValidOutputNodes = () => {
     const { ui } = this.stores;
-    if (!this.getNodeCount("Relay") && !this.getNodeCount("Data Storage")) {
+    if (!this.hasRelay() && !this.hasDataStorage()) {
       ui.alert("Program must contain a Relay or Data Storage node before it can be run.", "No Program Output");
       return false;
-    } else if (!this.getNodeCount("Relay") &&
+    } else if (!this.hasRelay() &&
                 this.programEditor.nodes.filter(n => (n.name === "Data Storage" && n.inputs.size <= 1)).length) {
       ui.alert("Data Storage node needs data inputs before program can be run.", "Missing Data Storage Inputs");
       return false;
     }
     return true;
   }
 
-  private runProgram = () => {
+  private prepareToRunProgram = () => {
+    const dialogPrompt = this.hasDataStorage()
+                          ? "Save dataset as"
+                          : "Save program as";
+    const baseTitle = this.getDatasetName() || this.context?.title || "program";
+    const programTitle = `${baseTitle}_${getLocalTimeStamp(Date.now())}`;
+    this.stores.ui.prompt(dialogPrompt, programTitle, "Run Program")
+    .then((title: string) => {
+      this.runProgram(title);
+    });
+  }
```

Comment:
If the user changes the filename, we get into an odd state.  The user-modified filename does not appear to be used and the dataset graph appears to have the wrong inputs.